### PR TITLE
Add sell function and tests

### LINKED ISSUE #10: Sell / Burn Functionality
- [ ] User should be able to sell the NFT back to the contract, which burns it upon sale
- [ ] The contract should take a 7% fee to the creator upon sale
- [ ] The contract should take a 3% fee to showtime upon sale
- [ ] The last NFT in circulation should not be burnable/sellable, because the contract will not have any escrow funds and the "price" will be assumed to be zero

### PR DESCRIPTION
**Pull request by garyghayrat**
closes #10

### CODE REVIEW

File: src/CreatorToken.sol [hunk 1]
```diff
@@ -4,13 +4,22 @@ pragma solidity 0.8.20;
...
+  error CreatorToken__CallerIsNotCreator(address _creator, address _caller);
+  error CreatorToken__CallerIsNotAdmin(address _admin, address _caller);
+  error CreatorToken__CallerIsNotCreatorOrAdmin(address _creator, address _admin, address _caller);
+  error CreatorToken__AddressZeroNotAllowed();
+  error CreatorToken__CallerIsNotOwner(uint256 _tokenId, address _owner, address _caller);
+  error CreatorToken__MinAcceptedPriceExceeded(uint256 _price, uint256 _minAcceptedPrice);
+  error CreatorToken__LastTokenCannotBeSold(uint256 _circulatingSupply);
 
   uint256 public lastId;
+  uint256 public circulatingSupply;
```
Comment (apbendi) [inline src/CreatorToken.sol:22]: Unless you had a specific reason for choosing this name, I think we should consider using `totalSupply` to harmonize with ERC20 instead.

File: src/CreatorToken.sol [hunk 2]
```diff
@@ -64,6 +111,35 @@ contract CreatorToken is ERC721 {
...
+    uint256 _tokenPrice = _temporaryGetNextTokenPrice();
+    (uint256 _creatorFee, uint256 _adminFee) = calculateFees(_tokenPrice);
+    uint256 _totalPrice = _tokenPrice - _creatorFee - _adminFee;
+
+    if (_totalPrice < _minAcceptedPrice) {
+      revert CreatorToken__MinAcceptedPriceExceeded(_totalPrice, _minAcceptedPrice);
+    }
+
+    transferFrom(msg.sender, address(this), _tokenId);
+    _burn(_tokenId);
```
Comment (apbendi) [inline src/CreatorToken.sol:129]: It looks like we don't decrement the circulating supply when we burn this. We should probably pull it into a small internal helper method `_burnAndDecrement` and do that. In addition to adding this, we should make sure the sell tests catch it. Thanks!

File: src/CreatorToken.sol [hunk 3]
```diff
@@ -64,6 +111,35 @@ contract CreatorToken is ERC721 {
...
+    emit Sold(msg.sender, _tokenId, _tokenPrice, _creatorFee, _adminFee);
+
+    payToken.safeTransfer(creator, _creatorFee);
+    payToken.safeTransfer(msg.sender, _totalPrice);
```
Comment (apbendi) [inline src/CreatorToken.sol:133]: This looks like a bug, right? sender should get `_tokenPrice`, not `_totalPrice`. Why don't our tests catch it?

**Review by apbendi — COMMENTED** [review]
Some initial notes @garyghayrat

File: src/CreatorToken.sol (hunk 3 above)
Comment (garyghayrat) [reply src/CreatorToken.sol:135]: I think the confusion is around the naming. 

I'm assuming that `_temporaryGetNextTokenPrice()` will always return `_tokenPrice`

When buying the `_totalPrice` the user pays is `_tokenPrice + _creatorFee + _adminFee;`
When selling the `_totalPrice` that the user gets is `_tokenPrice - _creatorFee - _adminFee;`

Let me know if the assumption is wrong or the naming still doesn't make sense.

File: src/CreatorToken.sol (hunk 1 above)
Comment (garyghayrat) [reply src/CreatorToken.sol]: works for me, updated in https://github.com/showtime-xyz/creator-tokens/pull/13/commits/3ad5836cd07b5ec29ed6f04b6058fde8a08e27a2

File: src/CreatorToken.sol (hunk 2 above)
Comment (garyghayrat) [reply src/CreatorToken.sol]: nice catch, fixed here: https://github.com/showtime-xyz/creator-tokens/pull/13/commits/3ad5836cd07b5ec29ed6f04b6058fde8a08e27a2

**Review by apbendi — APPROVED** [review]
Approved after feedback shared while pairing. Nice job @garyghayrat!

**Pull request merged**